### ISSUE TITLE
Investigate about AWS billing API endpoints

### ISSUE THREAD

**Issue by pavlov-tony** [pavlov-tony/xproject#3]
Acceptance criteria:
- Find out all endpoints that are responsible for Billing and payments.
- Figure out the availability of GRPC API.

**Comment by qezz** [comment]
## The API

AWS provides 2 ways to access their API:
1. Query API — it's like the REST, but not quite. There are different Query Requests for different services, for example: [for EC2](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/Query-Requests.html), [for Billing and Cost Management](https://docs.aws.amazon.com/awsaccountbilling/latest/aboutv2/using-pelong.html) 
2. AWS SDK libraries for a number of languages (Go among them). More about SDK for Go: https://aws.amazon.com/sdk-for-go/

Note that there is a [pricing policy for AWS Cost Management API usage](https://aws.amazon.com/aws-cost-management/pricing/). It turns out that the _AWS Cost Explorer Service_ and _AWS Budgets_ are not free to use ([see here](https://aws.amazon.com/aws-cost-management/pricing/)), but _AWS Price List Service_ seems free.

Are we going to use some of these?

**Comment by pavlov-tony** [comment]
@qezz, thanks for the detailed info about pricing policy. That information is very important for our future decisions. 
> Are we going to use some of these?

Definitely, we will use Go SDK! And we should try to use free queries. So let's try to understand what AWS Price List Service provides.
And well done Sergey!

**Comment by qezz** [comment]
## Steps to use the SDK:

### 1. Get SDK for Go
To get SDK and all deps:
```go get -u github.com/aws/aws-sdk-go/...```

(see [aws docs](https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/setting-up.html#install-go-sdk))

### 2. Get AWS access keys 

You should receive an __access key ID__ and __secret access key__ to sign requests.

Detailed instructions [here](https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/setting-up.html#get-aws-credentials)

### 3. Provide your credentials

You should provide your credentials in one of possible ways:

1. Shared Credentials File — `~/.aws/credentials`
2. Environment Variables — a bunch of
3. <strike>Hard-code them</strike> — no, you mustn't

More, as always, on aws docs — [here](https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/configuring-sdk.html#specifying-credentials)

### 4. You are great

Since now, you are allowed to write smth like this:

```go
sess := session.Must(session.NewSessionWithOptions(session.Options{
	SharedConfigState: session.SharedConfigEnable,
}))
```

[More about session API](https://docs.aws.amazon.com/sdk-for-go/api/aws/session/)

**Comment by qezz** [comment]
Below is a small experimental app which calls to Cost Explorer API

__Note: I'm not sure it's free, so be ready to pay $0.01 per request__

To enable Cost Explorer API you should allow it in your billing console: https://console.aws.amazon.com/billing/home#/

* More about access: https://docs.aws.amazon.com/awsaccountbilling/latest/aboutv2/cost-explorer-signup.html
* More about Cost Explorer (Go) SDK API: https://docs.aws.amazon.com/sdk-for-go/api/service/costexplorer/#GroupDefinition

<details>
  <summary>Example request (click to expand)</summary>

```go
package main

import (
	"fmt"
	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/costexplorer"
)

func main() {
	sess := session.Must(session.NewSessionWithOptions(session.Options{
		SharedConfigState: session.SharedConfigEnable,
	}))

	ceRegion := sess.Config.Region
	ce := costexplorer.New(sess, &aws.Config{Region: ceRegion})

	// Required fileds: Granularity, Metrics, TimePeriod

	// TimePeriod. startDate inclusive, endDate exclusive
	// Timestamp format is "YYYY-MM-DD"
	startDate := "2018-01-01"
	endDate := "2018-05-10"

	// Granularity: MONTHLY or DAILY
	granularity := "MONTHLY"

	// Selected metrics
	// Possible values: BlendedCost, UnblendedCost, and UsageQuantity
	usageqty := "UnblendedCost"
	selectedMetrics := []*string{
		&usageqty,
	}

	response, err := ce.GetCostAndUsage(&costexplorer.GetCostAndUsageInput{
		Granularity: &granularity,
		Metrics:     selectedMetrics,
		TimePeriod: &costexplorer.DateInterval{
			Start: &startDate,
			End:   &endDate,
		},
	})
	if err != nil {
		fmt.Println("error:", err)
		return
	}
	fmt.Println("output:", response)
}
```

</details>

**Comment by qezz** [comment]
## Price List API Usage Example 

Price List API is noted as free.

Requires `"us-east-1"` as a Region parameter

The code is rather noisy, I'll prettify it as soon as I figure out how to do this

<details><summary>Price List Example App</summary>

```go
package main

import (
	"fmt"
	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/pricing"
	"strconv"
)

// Trying Price List API
// Note: MaxResults is set to 1 to minimize output
func main() {
	sess := session.Must(session.NewSessionWithOptions(session.Options{
		SharedConfigState: session.SharedConfigEnable,
	}))

	pr := pricing.New(sess, aws.NewConfig().WithRegion("us-east-1"))

	// Get all services and their service codes
	{
		ds := pricing.DescribeServicesInput{
			MaxResults: aws.Int64(1),
		}
		response, err := pr.DescribeServices(&ds)
		if err != nil {
			fmt.Println("error:", err)
			return
		}
		fmt.Println("All Services:", response)
	}

	// Get AWSLambda price list
	{
		prods := pricing.GetProductsInput{
			MaxResults:  aws.Int64(1),
			ServiceCode: aws.String("AWSLambda"),
		}
		response, err := pr.GetProducts(&prods)
		if err != nil {
			fmt.Println("error:", err)
			return
		}
		fmt.Println("AWSLambda services:", response)
		// we should get `pricePerUnit`
		m := response.PriceList[0]["terms"].(map[string]interface{})

		// I don't know how to prettify these `.(map[string]interface{})`
		shortDesc := m["OnDemand"].(map[string]interface{})
		shortDesc = shortDesc["2MY95AYMSPD6MQ8Y.JRTCKXETXF"].(map[string]interface{})
		shortDesc = shortDesc["priceDimensions"].(map[string]interface{})
		shortDesc = shortDesc["2MY95AYMSPD6MQ8Y.JRTCKXETXF.6YS6EN2CT7"].(map[string]interface{})
		pricePU := shortDesc["pricePerUnit"].(map[string]interface{})
		ppu, err := strconv.ParseFloat(pricePU["USD"].(string), 64)
		if err != nil {
			ppu = 0.0
		}

		fmt.Println("\nShort Desc:")
		fmt.Println("\tService:", shortDesc["description"])
		fmt.Printf("\tPrice per million: $%v", ppu*1000000.0)
	}
}

```

</details>

<details><summary>Output</summary>

```
All Services: {
  FormatVersion: "aws_v1",
  NextToken: "mdP2mt5IIVX3adazwyiwag==:hgWthNqYGCnjkEhOazyfEgIEdsrfmY5phqXSRkp9IjjeP2gcdknd78dvNpR39wOvqcKciRDq2hW61zjwIyDJvsQOP24Ug/skahbhT3oQ2tv5QZRgqyolISyLsfb8FBukPydOG2bFcS00rtmmrYmUcQbT24GgEC/cRjL5RrfXBBKEl2rGdpes/p49yYXWmcU1",
  Services: [{
      AttributeNames: [
        "productFamily",
        "servicecode",
        "groupDescription",
        "termType",
        "usagetype",
        "locationType",
        "location"
      ],
      ServiceCode: "AWSBudgets"
    }]
}
AWSLambda services: {
  FormatVersion: "aws_v1",
  NextToken: "nAPpZMcu0uUCoqy6QHI6aQ==:NnHcbsw58AGBwXfPvdWlUUNgKRFMZESLfkJq397QM2q7FfFEdiYD6gEqZ/Jm1/kbLZcH7mmowgNx18Xrku94kwIuU5AxItWHkhpZlCmN1KTTckGEk9ycZxhveTN253FiR6Z/sv+n+JUh+WuhatQQUqNTNhgUAhvq3POz89TJnZbxp13TxAVx47amnjZcF+XN",
  PriceList: [{
      publicationDate: "2018-04-19T23:51:08Z",
      product: map[productFamily:Serverless attributes:map[locationType:AWS Region location:US West (N. California) servicename:AWS Lambda operation: group:AWS-Lambda-Edge-Requests servicecode:AWSLambda groupDescription:Invocation call for a Lambda function usagetype:USW1-Lambda-Edge-Request] sku:2MY95AYMSPD6MQ8Y],
      serviceCode: "AWSLambda",
      terms: map[OnDemand:map[2MY95AYMSPD6MQ8Y.JRTCKXETXF:map[sku:2MY95AYMSPD6MQ8Y effectiveDate:2018-04-01T00:00:00Z offerTermCode:JRTCKXETXF termAttributes:map[] priceDimensions:map[2MY95AYMSPD6MQ8Y.JRTCKXETXF.6YS6EN2CT7:map[unit:Request endRange:Inf description:AWS Lambda Edge - Total Requests - US West (N. California) appliesTo:[] rateCode:2MY95AYMSPD6MQ8Y.JRTCKXETXF.6YS6EN2CT7 beginRange:0 pricePerUnit:map[USD:0.0000006000]]]]]],
      version: "20180419235108"
    }]
}

Short Desc:
	Service: AWS Lambda Edge - Total Requests - US West (N. California)
	Price per million: $0.6

```

<details>

**Comment by qezz** [comment]
## Cost and Usage Report service

**TL;DR**: Hourly (or daily) collects your usage data and puts it in S3 as a CSV file. The reports are free of charge, but standard Amazon S3 storage rates apply.

More details about [Cost and Usage Report service](https://docs.aws.amazon.com/awsaccountbilling/latest/aboutv2/billing-reports-costusage.html).

We should be able to take them from S3 every hour or so. There is also a Free Tier limit for S3, so we can use it for a while without charges.

Step by step guide about how to enable Cost and Usage Report: https://docs.aws.amazon.com/awsaccountbilling/latest/aboutv2/billing-reports-gettingstarted.html

**Important**: It takes at least 24h to enable Cost and Usage Report service. 

An example will be added as soon as service become available for me.

**Comment by strelnikov-sv** [comment]
**Maybe better to use v2 API (But this version of preview 3)?** 
https://github.com/aws/aws-sdk-go-v2
Instruction: https://godoc.org/github.com/aws/aws-sdk-go-v2/aws

About what's new here: https://github.com/aws/aws-sdk-go-v2/blob/master/CHANGELOG.md#release-v200-preview1-2017-12-21

**Comment by qezz** [comment]
@strelnikov-sv I thought about it, but the [documentation](https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/welcome.html) and entire [API reference](https://docs.aws.amazon.com/sdk-for-go/api/) are based on v1 (https://github.com/aws/aws-sdk-go) version.

**Comment by strelnikov-sv** [comment]
Foreign project visualization of billing AWS: https://github.com/PriceBoardIn/aws-elk-billing
